str() to accept optional length

Birch-san opened the pull request:
Fixes https://github.com/iovisor/bpftrace/issues/199. For _small_ strings, at least.

```bash
sudo bpftrace -e 'tracepoint:syscalls:sys_enter_write /pid == 23506/
{
  printf("<%s>\n", str(args->buf, args->count)); 
}'
# type pwd into terminal 23506
<p>
<w>
<d>
# press enter in terminal 23506
<
>
</home/anon
>
# notice that this string got truncated; we don't allocate sufficiently large strings (by default)
<anon@anon-VirtualBox:

# meanwhile, here's what terminal 23506 looks like:
anon@anon-VirtualBox:~$ pwd
/home/anon
anon@anon-VirtualBox:~$ 
```

By default, the largest string it can support is 64 bytes. We can tune this with the environment variable `BPFTRACE_STRLEN` — which fixes the string truncation in this particular situation:

```bash
sudo BPFTRACE_STRLEN=240 bpftrace -e 'tracepoint:syscalls:sys_enter_write /pid == 25783/ { printf("<%s>\n", str(args->buf, args->count)); }'
<
>
</home/anon
>
<anon@anon-VirtualBox:~$ >
```

Caveat: str() allocates strings on the BPF stack, so we reach the 512-byte limit very quickly (especially since we pay the toll again when printf() composes a perf event output buffer).

I explored the possibility of storing the strings off-stack. The main problem there is that probe_read_str() will only write upon on-stack memory. I have ideas for workarounds (and some code progress towards those workarounds), but they're too big to be included in the scope of this pull request.

Birch-san commented:
Here's some code that I built to explore the possibility of storing strings off-stack:

https://github.com/iovisor/bpftrace/compare/bd038f9110009f282c657c393fae60fdb7dd50ba...Birch-san:c5088c9a5d731165e1005a6a800e611bb63f3dc3

It introduces (in codegen) a "StrCall" specialization of "Call".  
StrCall has a `PERCPU_ARRAY` BPF map, to hold large data off-stack.

We cannot use probe_read_str() to output directly into the BPF map, but maybe we could allocate a small on-stack buffer, and ferry bytes (one buffer at a time) from BPF stack into BPF map.

Next, printf() poses a few problems:

- composes the perf event output struct on the BPF stack
- doesn't use runtime information to decide length of strings inside the perf event output struct
  - composes a fixed-size structure
  - determines size of structure during semantic analysis

I _think_ PerfEventOutput can transfer data from _BPF map_ to userspace (that's what "join" does), so we could compose the perf event output struct inside a BPF map.

We may find that we want a dynamically-sized perf event output structure. Like, imagine we see that args->count is just 3 bytes. We wouldn't want to send MAX_STRLEN bytes to userspace, if we know at runtime a smaller, more specific amount.

Here's some (horrifying) code I wrote to make print decide the size of its MEMCPY based on runtime information rather than semantic analysis:

https://github.com/iovisor/bpftrace/compare/master...Birch-san:6c1f19a3a4c3cf253aa23db79f5f637b6823e5b2

Notice how expr2_ is used to pass down a strlen at runtime, instead of using the strlen provided during semantic analysis.  
And notice how the str call uses a parameterised AllocInst, so instead of `isArrayTy()`, its expr_ type has `isPointerTy()`. (it's a pointer to dynamically-allocated memory)

I reached a dead-end here because printf() allocates a buffer based on an LLVM StructType. It can store strings in ArrayType members, but you need to declare the size of the ArrayType up front. You cannot use an LLVM Value to dynamically size it.  
The workaround here would be to stop using StructType and instead use a basic AllocInst. These support dynamic sizes based on LLVM Value. Or even better: we can use off-stack memory (a BPF map) instead of doing an allocation inside printf().

====

In conclusion: there _may_ be a path to supporting arbitrarily-large strings. Bonus points if we can also make the perf event output buffer dynamically-sized, so we can transfer the smallest message possible (instead of our maximum buffer size). A big advantage of dynamic sizing is that the user wouldn't have to tune their maximum buffer size for the sake of performance.

brendangregg commented:
Thanks for this work, and that's a good investigation for future support of larger strings -- for which we probably will need a string map that can be implemented similar to join's needs_join_map_ etc. I'd create a new ticket for this "str() to accept large lengths" and refer to this PR comment.

As for this PR:

- The USAGE message is getting too big and wide (I avoid >80 char width). I'd reduce it to just:

```
ENVIRONMENT:
    BPFTRACE_STRLEN    [default: 64] bytes on BPF stack by str().
```

(which needs an extra std::endl at the end), and then put a section in the reference guide with a longer explanation. Maybe under the first Usage section, a new "7. Environment Variables". The docs should be as short as possible.

- If a user sets strlen to be something large, I imagine they'll get a compiler error. I'd put a check when the environment variable is read for some sane maximum -- eg, 400 bytes -- and then print a detailed error message if that is exceeded. Something like:

```
ERROR: str() length %d exceeds the current maximum of 400 bytes.
This limitation is because strings are currently stored on the 512 byte BPF stack.
A future enhancement will remove this limitation.
```

Birch-san commented:
> I'd create a new ticket for this "str() to accept large lengths"

Done: https://github.com/iovisor/bpftrace/issues/305

> USAGE message is getting too big and wide

I've committed the following docs changes:

- shortened the -h message
- changed method signature in `README.md`
- detailed str() and env in `reference_guide.md`
- added learnings to `internals_development.md`

> check when the environment variable is read for some sane maximum

Experimentally, the largest buffer I've fit into the stack (with a str() and a print()) is 240 bytes. I've added a friendly 200-byte limit + error message that will protect the user in the average case.

If they have a deep stack, or multiple calls to str() in their program, then the parameter validation won't save them. But the message from the verifier (stack limit exceeded, large on-stack variables) is pretty instructive anyway.

brendangregg commented:
Excellent, and thanks for all the additions to the internals development guide!
